File: ChatSupport/src/main/java/com/vantagecircle/chatapp/activity/LoginActivity.java
package com.vantagecircle.chatapp.activity;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatEditText;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.vantagecircle.chatapp.R;
import com.vantagecircle.chatapp.services.SupportService;
import com.vantagecircle.chatapp.core.model.DataModel;
import com.vantagecircle.chatapp.core.GetDataHandler;
import com.vantagecircle.chatapp.core.interfaceC.ResultInterface;
import com.vantagecircle.chatapp.core.interfaceC.ValueInterface;
import com.vantagecircle.chatapp.utils.Constants;
import com.vantagecircle.chatapp.core.AuthHandler;
import com.vantagecircle.chatapp.utils.UpdateKeyUtils;
import com.vantagecircle.chatapp.model.UserM;
import com.vantagecircle.chatapp.pref.SharedPrefM;
import com.vantagecircle.chatapp.utils.ToolsUtils;

/**
 * Created by bapidas on 10/07/17.
 */

public class LoginActivity extends AppCompatActivity {
    private static final String TAG = LoginActivity.class.getSimpleName();
    private Activity activity;
    private Context mContext;
    private EditText usernameEdit, passwordEdit;
    private Button btnLogin, btnSignup;
    private ProgressDialog progressDialog;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activity = this;
        mContext = getApplicationContext();
        setContentView(R.layout.activity_login);
        initView();
        initListener();
    }

    private void initView() {
        usernameEdit = (AppCompatEditText) findViewById(R.id.usernameEdit);
        passwordEdit = (AppCompatEditText) findViewById(R.id.passwordEdit);
        btnLogin = (Button) findViewById(R.id.btnLogin);
        btnSignup = (Button) findViewById(R.id.btnSignup);
    }

    private void initListener() {
        btnLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (usernameEdit.getText().toString().length() == 0 &&
                        passwordEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Username and password is cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else if (usernameEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Username cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else if (passwordEdit.getText().toString().length() == 0) {
                    Toast.makeText(mContext, "Password cannot be blank",
                            Toast.LENGTH_SHORT).show();
                } else {
                    if (ToolsUtils.isNetworkAvailable(mContext)) {
                        ToolsUtils.hideKeyboard(activity);
                        progressDialog = new ProgressDialog(activity);
                        progressDialog.setMessage("Authenticating user");
                        progressDialog.show();
                        userLogin();
                    } else {
                        Toast.makeText(mContext, "Please check your internet connection",
                                Toast.LENGTH_SHORT).show();
                    }
                }
            }
        });

        btnSignup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(activity, SignupActivity.class);
                startActivity(intent);
            }
        });
    }

    private void userLogin() {
        String username = usernameEdit.getText().toString();
        String password = passwordEdit.getText().toString();

        AuthHandler authHandler = new AuthHandler();
        authHandler.setFirebaseAuth(SupportService.getAuthInstance());
        authHandler.performLogin(username, password, new ResultInterface() {
            @Override
            public void onSuccess(String t) {
                progressDialog.setMessage("Authentication success");
                UpdateKeyUtils.updateTokenToServer(new SharedPrefM(mContext).getString(Constants.FIREBASE_TOKEN));
                getData();
            }

            @Override
            public void onFail(String e) {
                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }
                Toast.makeText(mContext, e, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void getData(){
        GetDataHandler getDataHandler = new GetDataHandler();
        getDataHandler.setDataReference(SupportService.getUserReference().child(SupportService.getUserInstance().getUid()));
        getDataHandler.setSingleValueEventListener(new ValueInterface() {
            @Override
            public void onDataSuccess(DataModel dataModel) {
                UserM userM = dataModel.getDataSnapshot().getValue(UserM.class);
                if (userM != null) {
                    SupportService.init(getApplicationContext(), userM);
                    if (progressDialog != null && progressDialog.isShowing()) {
                        progressDialog.dismiss();
                    }
                    if (userM.getUserType().equals(Constants._ADMIN)) {
                        SupportService.getAuthInstance().signOut();
                        Toast.makeText(mContext, "User invalid use another account",
                                Toast.LENGTH_SHORT).show();
                    } else {
                        Intent intent = new Intent(LoginActivity.this, UserActivity.class);
                        startActivity(intent);
                        finish();
                    }
                } else {
                    SupportService.getAuthInstance().signOut();
                    if (progressDialog != null && progressDialog.isShowing()) {
                        progressDialog.dismiss();
                    }
                    Toast.makeText(mContext, "User data fetch error try again",
                            Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onDataCancelled(DataModel databaseError) {
                SupportService.getAuthInstance().signOut();
                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }
                Toast.makeText(mContext, databaseError.getDatabaseError().getMessage(),
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onResume() {
        Log.d(TAG, "onResume");
        super.onResume();
    }

    @Override
    protected void onPause() {
        Log.d(TAG, "onPause");
        super.onPause();
    }

    @Override
    protected void onStart() {
        Log.d(TAG, "onStart");
        super.onStart();
    }

    @Override
    protected void onStop() {
        Log.d(TAG, "onStop");
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "onDestroy");
        super.onDestroy();
    }
}

File: app/src/main/java/com/bapidas/chattingapp/ui/main/LoginActivity.kt
package com.bapidas.chattingapp.ui.main

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatEditText
import com.bapidas.chattingapp.ChatApplication
import com.bapidas.chattingapp.R
import com.bapidas.chattingapp.data.core.AuthHandler
import com.bapidas.chattingapp.data.core.GetDataHandler
import com.bapidas.chattingapp.data.core.callbacks.ResultInterface
import com.bapidas.chattingapp.data.core.callbacks.ValueInterface
import com.bapidas.chattingapp.data.core.model.DataModel
import com.bapidas.chattingapp.data.model.User
import com.bapidas.chattingapp.data.pref.SharedPrefHelper
import com.bapidas.chattingapp.utils.Constants
import com.bapidas.chattingapp.utils.ToolsUtils.hideKeyboard
import com.bapidas.chattingapp.utils.ToolsUtils.isNetworkAvailable
import com.bapidas.chattingapp.utils.UpdateKeyUtils.updateTokenToServer

/**
 * Created by bapidas on 10/07/17.
 */
class LoginActivity : AppCompatActivity() {
    lateinit var usernameEdit: EditText
    lateinit var passwordEdit: EditText
    lateinit var btnLogin: Button
    lateinit var btnSignup: Button
    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        initView()
        initListener()
    }

    private fun initView() {
        usernameEdit = findViewById<View>(R.id.usernameEdit) as AppCompatEditText
        passwordEdit = findViewById<View>(R.id.passwordEdit) as AppCompatEditText
        btnLogin = findViewById<View>(R.id.btnLogin) as Button
        btnSignup = findViewById<View>(R.id.btnSignup) as Button
    }

    private fun initListener() {
        btnLogin.setOnClickListener {
            if (usernameEdit.text.toString().isEmpty() &&
                    passwordEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Username and password is cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else if (usernameEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Username cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else if (passwordEdit.text.toString().isEmpty()) {
                Toast.makeText(this, "Password cannot be blank",
                        Toast.LENGTH_SHORT).show()
            } else {
                if (isNetworkAvailable(this)) {
                    hideKeyboard(this)
                    progressDialog = ProgressDialog(this)
                    progressDialog?.setMessage("Authenticating user")
                    progressDialog?.show()
                    userLogin()
                } else {
                    Toast.makeText(this, "Please check your internet connection",
                            Toast.LENGTH_SHORT).show()
                }
            }
        }
        btnSignup.setOnClickListener {
            val intent = Intent(this, SignupActivity::class.java)
            startActivity(intent)
        }
    }

    private fun userLogin() {
        val username = usernameEdit.text.toString()
        val password = passwordEdit.text.toString()
        val authHandler = AuthHandler()
        authHandler.firebaseAuth = ChatApplication.applicationContext().authInstance
        authHandler.performLogin(username, password, object : ResultInterface {
            override fun onSuccess(t: String) {
                progressDialog?.setMessage("Authentication success")
                updateTokenToServer(SharedPrefHelper(this@LoginActivity)
                        .getString(Constants.FIREBASE_TOKEN))
                data
            }

            override fun onFail(e: String) {
                if (progressDialog != null && progressDialog?.isShowing == true) {
                    progressDialog?.dismiss()
                }
                Toast.makeText(this@LoginActivity, e, Toast.LENGTH_SHORT).show()
            }
        })
    }

    private val data: Unit
        get() {
            val getDataHandler = GetDataHandler()
            val ref = ChatApplication.applicationContext().userReference
                    .child(ChatApplication.applicationContext().userInstance?.uid.orEmpty())
            getDataHandler.setSingleValueEventListener(ref, object : ValueInterface {
                override fun onDataSuccess(dataModel: DataModel) {
                    val userM = dataModel.dataSnapshot?.getValue(User::class.java)
                    if (userM != null) {
                        ChatApplication.applicationContext().user = userM
                        if (progressDialog != null && progressDialog?.isShowing == true) {
                            progressDialog?.dismiss()
                        }
                        if (userM.userType == Constants.ADMIN) {
                            ChatApplication.applicationContext().authInstance.signOut()
                            Toast.makeText(this@LoginActivity,
                                    "User invalid use another account",
                                    Toast.LENGTH_SHORT).show()
                        } else {
                            val intent = Intent(this@LoginActivity, UserActivity::class.java)
                            startActivity(intent)
                            finish()
                        }
                    } else {
                        ChatApplication.applicationContext().authInstance.signOut()
                        if (progressDialog != null && progressDialog?.isShowing == true) {
                            progressDialog?.dismiss()
                        }
                        Toast.makeText(this@LoginActivity, "User data fetch error try again",
                                Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onDataCancelled(dataModel: DataModel) {
                    ChatApplication.applicationContext().authInstance.signOut()
                    if (progressDialog != null && progressDialog?.isShowing == true) {
                        progressDialog?.dismiss()
                    }
                    Toast.makeText(this@LoginActivity, dataModel.databaseError?.message,
                            Toast.LENGTH_SHORT).show()
                }
            })
        }
}